components/supervisor/pkg/ports/ports.go

```
@@ -0,0 +1,575 @@
+			mp = &managedPort{}
+			state[port] = mp
+		}
+
+		mp.LocalhostPort = port
+		mp.Served = true
+
+		exposedGlobalPort := mp.GlobalPort
+		if served.BoundToLocalhost {
+			proxy, exists := pm.proxies[port]
+			if exists {
+				mp.GlobalPort = proxy.proxyPort
+			} else {
+				mp.GlobalPort = 0
+			}
+		} else {
+			// we don't need a proxy - the port is globally bound
+			mp.GlobalPort = port
+		}
+
+		if mp.GlobalPort == 0 || (mp.Exposed && mp.GlobalPort == exposedGlobalPort) {
+			continue
+		}
+
+		var public bool
+		_, configured := pm.configs.Get(mp.LocalhostPort)
+		if mp.Exposed || configured {
+			public = mp.Public
+		} else {
+			config, exists := pm.configs.GetFromRange(port)
+			public = exists && config.Visibility != "private"
+		}
+
+		err := pm.E.Expose(ctx, mp.LocalhostPort, mp.GlobalPort, public)
+		if err != nil {
+			log.WithError(err).WithField("port", *mp).Warn("cannot auto-expose port")
+			continue
+		}
+		log.WithField("port", *mp).Warn("auto-expose port")
+	}
+	return state
+}
+
+func getOnExposed(config *gitpod.PortConfig, port uint32) api.PortsStatus_ExposedPortInfo_OnPortExposed {
+	if config == nil {
+		// anything above 32767 seems odd (e.g. used by language servers)
+		unusualRange := !(0 < port && port < 32767)
+		wellKnown := port <= 10000
+		if unusualRange || !wellKnown {
+			return api.PortsStatus_ExposedPortInfo_ignore
+		}
+		return api.PortsStatus_ExposedPortInfo_notify_private
+	}
+	if config.OnOpen == "ignore" {
+		return api.PortsStatus_ExposedPortInfo_ignore
+	}
+	if config.OnOpen == "open-browser" {
+		return api.PortsStatus_ExposedPortInfo_open_browser
+	}
+	if config.OnOpen == "open-preview" {
+		return api.PortsStatus_ExposedPortInfo_open_preview
+	}
+	return api.PortsStatus_ExposedPortInfo_notify
+}
+
+func (pm *Manager) boundInternally(port uint32) bool {
+	_, exists := pm.internal[port]
+	return exists
+}
+
+// Expose exposes a port
+func (pm *Manager) Expose(port uint32, targetPort uint32) string {
```

Why not return a proper error?